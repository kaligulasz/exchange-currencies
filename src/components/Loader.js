import React from 'react';
import styled from 'styled-components';
import { Spinner5 } from 'styled-icons/icomoon/Spinner5';

const Background = styled.div`
  background: ${props => props.theme.color.primaryGradient};
  position: absolute;
  height: 100%;
  width: 100%;
  display: flex;
  justify-content: center;
  z-index: 1;
`;

const StyledSpinner = styled(Spinner5)`
  width: 3.5rem;
  animation: spin 1.2s linear infinite;
  
  @keyframes spin {
  0% {
    transform: rotate(0deg);
  }
  100% {
    transform: rotate(360deg);
  }
}
`;

const Loader = () => (
  <Background>
    <StyledSpinner />
  </Background>
);

export default Loader;
