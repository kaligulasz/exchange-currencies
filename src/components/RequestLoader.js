import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';

// Components
import Loader from './Loader';

const Wrapper = styled.div`
  position: relative;
`;

const RequestLoader = ({ status, children }) => (
  <Wrapper>
    {status === 'initial'
      && <Loader />
    }

    {(status === 'successful' || status === 'initial')
      && children
    }

    {status === 'error'
      && 'error'
    }
  </Wrapper>
);

RequestLoader.propTypes = {
  status: PropTypes.string.isRequired,
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node,
  ]).isRequired,
};

export default RequestLoader;
