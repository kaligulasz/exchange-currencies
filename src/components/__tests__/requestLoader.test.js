import theme from '../../common/themeVariables';

import RequestLoader from '../RequestLoader';
import Loader from '../Loader';
import ErrorMessage from '../ErrorMessage';
import { Error } from 'styled-icons/boxicons-regular/Error';
import React from 'react';
import styled from 'styled-components';

const ErrorIcon = styled(Error)`
  margin-right: 0.5em;
`;

describe('<RequestLoader />', () => {
  it('renders loader component when status is initial', () => {
    const componentWithChildren = shallow(
      <RequestLoader status="initial">
        <div>example children</div>
      </RequestLoader>,
    );

    expect(componentWithChildren.contains(<Loader />)).toBe(true);
  });

  it('renders children when status is successful', () => {
    const componentWithChildren = shallow(
      <RequestLoader status="successful">
        <div>example children</div>
      </RequestLoader>,
    );

    expect(componentWithChildren.contains(<div>example children</div>)).toBe(true);
  });

  it('renders error component when status is failed', () => {
    const componentWithChildren = shallow(
      <RequestLoader status="failed">
        <div>example children</div>
      </RequestLoader>,
    );

    expect(componentWithChildren).toMatchSnapshot();
  });
});
