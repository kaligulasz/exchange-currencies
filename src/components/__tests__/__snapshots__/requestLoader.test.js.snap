// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RequestLoader /> renders error component when status is failed 1`] = `
<RequestLoader__Wrapper>
  <Message
    message="something went wrong"
  >
    <RequestLoader__ErrorIcon
      size="35"
    />
  </Message>
</RequestLoader__Wrapper>
`;
